src/data/simple_dmatrix.h

```diff
@@ -26,6 +28,116 @@ class SimpleDMatrix : public DMatrix {
+      size_t num_lines = batch.Size();
+#pragma omp parallel for schedule(static)
+      for (omp_ulong i = 0; i < static_cast<omp_ulong>(num_lines);
+        ++i) {  // NOLINT(*)
+        int tid = omp_get_thread_num();
+        auto line = batch.GetLine(i);
+        for (auto j = 0ull; j < line.Size(); j++) {
+          auto element = line.GetElement(j);
+          inferred_num_columns =
+              std::max(inferred_num_columns,
+                       static_cast<uint64_t>(element.column_idx + 1));
+          if (!common::CheckNAN(element.value) && element.value != missing) {
+            builder.AddBudget(element.row_idx, tid);
+          }
+        }
+      }
+      builder.InitStorage();
+
+      // Second pass over batch, placing elements in correct position
+#pragma omp parallel for schedule(static)
+      for (omp_ulong i = 0; i < static_cast<omp_ulong>(num_lines);
+        ++i) {  // NOLINT(*)
+        int tid = omp_get_thread_num();
+        auto line = batch.GetLine(i);
+        for (auto j = 0ull; j < line.Size(); j++) {
+          auto element = line.GetElement(j);
+          if (!common::CheckNAN(element.value) && element.value != missing) {
+            builder.Push(element.row_idx, Entry(element.column_idx, element.value),
+              tid);
+          }
+        }
+      }
+
+      // Append meta information if available
+      if (batch.Labels() != nullptr) {
+        auto& labels = mat.info.labels_.HostVector();
+        labels.insert(labels.end(), batch.Labels(), batch.Labels() + batch.Size());
+      }
+      if (batch.Weights() != nullptr) {
+        auto& weights = mat.info.weights_.HostVector();
+        weights.insert(weights.end(), batch.Weights(), batch.Weights() + batch.Size());
+      }
+      if (batch.Qid() != nullptr) {
+        qids.insert(qids.end(), batch.Qid(), batch.Qid() + batch.Size());
+        // get group
+        for (size_t i = 0; i < batch.Size(); ++i) {
+          const uint64_t cur_group_id = batch.Qid()[i];
+          if (last_group_id == default_max || last_group_id != cur_group_id) {
+            mat.info.group_ptr_.push_back(group_size);
+          }
+          last_group_id = cur_group_id;
+          ++group_size;
+        }
+      }
+    }
+
+    if (last_group_id != default_max) {
+      if (group_size > mat.info.group_ptr_.back()) {
+        mat.info.group_ptr_.push_back(group_size);
+      }
+    }
+
+    // Deal with empty rows/columns if necessary
+    if (adapter->NumColumns() == 0) {
+      mat.info.num_col_ = inferred_num_columns;
+    } else {
+      mat.info.num_col_ = adapter->NumColumns();
+    }
+
+    if (offset_vec.size() - 1 < adapter->NumRows() && adapter->NumRows() > 0) {
+      while (offset_vec.size() - 1 < adapter->NumRows()) {
+        offset_vec.emplace_back(offset_vec.back());
+      }
+      mat.info.num_row_ = adapter->NumRows();
+    } else {
+      mat.info.num_row_ = offset_vec.size() - 1;
+    }
+
+    mat.info.num_nonzero_ = offset_vec.back();
```

Comment:
Unless you are sure about there must be at least 1 element, don't use `front` `back`.